File: algorithms/javaalgo/src/test/java/programmers/level2/binarychangerepeat/BinaryScaleChangeRepeatTest.java
package programmers.level2.binarychangerepeat;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import org.junit.jupiter.api.Test;

public class BinaryScaleChangeRepeatTest {

    @Test
    void test() {
        String s1 = "110010101001";
        String s2 = "01110";
        String s3 = "1111111";

        assertAll(
            () -> assertThat(solution(s1)).isEqualTo(new int[]{3, 8}),
            () -> assertThat(solution(s2)).isEqualTo(new int[]{3, 3}),
            () -> assertThat(solution(s3)).isEqualTo(new int[]{4, 1})
        );
    }

    public int[] solution(String s) {
        int[] answer = new int[2];

        while (s.length() > 1) {
            int oneCount = 0;

            for (int i = 0; i < s.length(); i++) {
                if (s.charAt(i) == '0') {
                    answer[1]++;
                } else {
                    oneCount++;
                }
            }
            s = Integer.toBinaryString(oneCount);
            answer[0]++;
        }
        return answer;
    }
}

File: algorithms/javascriptalgo/programmers/level2/binaryScaleChangeRepeat.test.js
test("test", () => {
    let s1 = "110010101001";
    let s2 = "01110";
    let s3 = "1111111";

    expect(solution(s1)).toEqual([3, 8]);
    expect(solution(s2)).toEqual([3, 3]);
    expect(solution(s3)).toEqual([4, 1]);
});

function solution(s) {
    const answer = [0, 0];
    while (s.length > 1) {
        let oneCount = 0;

        for (let i = 0; i < s.length; i++) {
            if (s[i] === '0') {
                answer[1]++;
            } else {
                oneCount++;
            }
        }
        s = oneCount.toString(2);
        answer[0]++;
    }

    return answer;
}

function solution2(s) {
    var answer = [0, 0];
    while (s.length > 1) {
        answer[0]++;
        // /0/g 는 0인 숫자 전부를 의미, [] 0이 하나도 없을 경우 null 방지 대비
        answer[1] += (s.match(/0/g) || []).length;
        s = s.replace(/0/g, '').length.toString(2);
    }
    return answer;
}
